sdk/core/core/src/az_http_policy.c

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_http_pipeline.h>
+#include <az_http_policy.h>
+
+#include <_az_cfg.h>
+
+AZ_NODISCARD AZ_INLINE az_result az_http_pipeline_nextpolicy(
+    az_http_policy * const p_policies,
+    az_http_request_builder * const hrb,
+    az_mut_span const * const response) {
+
+  AZ_CONTRACT_ARG_NOT_NULL(p_policies);
+  AZ_CONTRACT_ARG_NOT_NULL(hrb);
+  AZ_CONTRACT_ARG_NOT_NULL(response);
+
+  // Transport Policy is the last policy in the pipeline
+  //  it returns without calling nextpolicy
+  if (&(p_policies[0]) == NULL || p_policies[0].pfnc_process == NULL) {
+    return AZ_ERROR_HTTP_PIPELINE_INVALID_POLICY;
+  }
+
+  return p_policies[0].pfnc_process(&(p_policies[1]), hrb, response);
+}
+
+static az_span const AZ_MS_CLIENT_REQUESTID = AZ_CONST_STR("x-ms-client-request-id");
+
+AZ_NODISCARD az_result az_http_pipeline_policy_uniquerequestid(
+    az_http_policy * const p_policies,
+    az_http_request_builder * const hrb,
+    az_mut_span const * const response) {
+  AZ_CONTRACT_ARG_NOT_NULL(p_policies);
+  AZ_CONTRACT_ARG_NOT_NULL(hrb);
+  AZ_CONTRACT_ARG_NOT_NULL(response);
+
+  // TODO - add a UUID create implementation
+  az_span const uniqueid = AZ_CONST_STR("123e4567-e89b-12d3-a456-426655440000");
+
+  // Append the Unique GUID into the headers
+  //  x-ms-client-request-id
+  AZ_RETURN_IF_FAILED(az_http_request_builder_append_header(hrb, AZ_MS_CLIENT_REQUESTID, uniqueid));
+
+  return az_http_pipeline_nextpolicy(p_policies, hrb, response);
+}
+
+AZ_NODISCARD az_result az_http_pipeline_policy_retry(
+    az_http_policy * const p_policies,
+    az_http_request_builder * const hrb,
+    az_mut_span const * const response) {
+  AZ_CONTRACT_ARG_NOT_NULL(p_policies);
+  AZ_CONTRACT_ARG_NOT_NULL(hrb);
+  AZ_CONTRACT_ARG_NOT_NULL(response);
+  // Retry logic
+  return az_http_pipeline_nextpolicy(p_policies, hrb, response);
+}
+
+AZ_NODISCARD az_result az_http_pipeline_policy_authentication(
+    az_http_policy * const p_policies,
+    az_http_request_builder * const hrb,
+    az_mut_span const * const response) {
+  AZ_CONTRACT_ARG_NOT_NULL(p_policies);
+  AZ_CONTRACT_ARG_NOT_NULL(hrb);
+  AZ_CONTRACT_ARG_NOT_NULL(response);
+  // Authentication logic
+  return az_http_pipeline_nextpolicy(p_policies, hrb, response);
+}
+
+AZ_NODISCARD az_result az_http_pipeline_policy_logging(
+    az_http_policy * const p_policies,
+    az_http_request_builder * const hrb,
+    az_mut_span const * const response) {
+  AZ_CONTRACT_ARG_NOT_NULL(p_policies);
+  AZ_CONTRACT_ARG_NOT_NULL(hrb);
+  AZ_CONTRACT_ARG_NOT_NULL(response);
+  // Authentication logic
+  return az_http_pipeline_nextpolicy(p_policies, hrb, response);
+}
+
+AZ_NODISCARD az_result az_http_pipeline_policy_bufferresponse(
+    az_http_policy * const p_policies,
+    az_http_request_builder * const hrb,
+    az_mut_span const * const response) {
```

Comment:
Don't pass `az_mut_span const` by pointer. I would say we should use

- `TYPE const param` for parameter that can be copied by value and have a size less than about 4 pointers.
- `TYPE const * const param` for parameters that bigger that 4 pointers or can't be copied by value.
- `TYPE * const param` for parameters that can be mutated/initialized.

Reply:
I would suggest to use `az_mut_span const response`.

Reply:
@sergey-shandar , then how would the code modify response's size?